### PR TITLE
Validating synapse artifacts from synapse configuration in gateway RestAPI

### PR DESCRIPTION
**Pull request by Sarangan0219**
Related issue: wso2/product-apim#8246

Here after fetching the corresponding synapse artifacts from DB, It has been validated with the deployed synapse artifacts in synapse configuration and return to the restAPI call.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 1]
```diff
@@ -18,24 +18,101 @@
...
+
 
 import javax.ws.rs.core.Response;
 
 public class ApiArtifactApiServiceImpl implements ApiArtifactApiService {
 
-    public Response apiArtifactGet(String apiName, String label, String apiId, MessageContext messageContext) {
+    private static final Log log = LogFactory.getLog(ApiArtifactApiServiceImpl.class);
+
+    private ApiMgtDAO apiMgtDAO = ApiMgtDAO.getInstance();
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:45]: We should not access DAO layer outside impl. Even in gateway profile, we might not have access to db.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 2]
```diff
@@ -18,24 +18,101 @@
...
+    public Response apiArtifactGet(String apiName, String version , String tenantDomain, MessageContext messageContext) {
 
         InMemoryAPIDeployer inMemoryApiDeployer = new InMemoryAPIDeployer();
+        if (tenantDomain == null){
+            tenantDomain =SUPER_TENAT_DOMAIN;
+        }
+        String apiId = null;
+        String label = null;
+        try {
+            apiId = apiMgtDAO.getGatewayAPIId(apiName,version,tenantDomain);
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:58]: use the artifact retriever to get the gateway API Id

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 3]
```diff
@@ -18,24 +18,101 @@
...
+            label = apiMgtDAO.getGatewayAPILabel(apiId);
+        } catch (APIManagementException e) {
+            log.error(e);
+        }
         GatewayAPIDTO gatewayAPIDTO = inMemoryApiDeployer.getAPIArtifact(apiId, label);
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:63]: What will happen if the artifact status is "remove"? We have to check for that before retrieving the artifacts

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 4]
```diff
@@ -18,24 +18,101 @@
...
+            try {
+                if (gatewayAPIDTO.getEndpointEntriesToBeAdd() != null ) {
+                    JSONArray endPointArray = new JSONArray();
+                    EndpointAdminServiceProxy endpointAdminServiceProxy = new EndpointAdminServiceProxy
+                            (gatewayAPIDTO.getTenantDomain());
+                    for (GatewayContentDTO gatewayEndpoint : gatewayAPIDTO.getEndpointEntriesToBeAdd()) {
+                        if (endpointAdminServiceProxy.isEndpointExist(gatewayEndpoint.getName())) {
+                            endPointArray.put(endpointAdminServiceProxy.getEndpoints(gatewayEndpoint.getName()));
+                        } else {
+                            log.error(gatewayEndpoint.getName() + " was not deployed in the gateway");
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:77]: Shall we add this details to the response? It is good to know which endpoints were not added.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 5]
```diff
@@ -18,24 +18,101 @@
...
+
+                if (gatewayAPIDTO.getLocalEntriesToBeAdd() != null) {
+                    JSONArray localEntryArray = new JSONArray();
+                    LocalEntryServiceProxy localEntryServiceProxy = new
+                            LocalEntryServiceProxy(gatewayAPIDTO.getTenantDomain());
+                    for (GatewayContentDTO localEntry : gatewayAPIDTO.getLocalEntriesToBeAdd()) {
+                        if (localEntryServiceProxy.isEntryExists(localEntry.getName())) {
+                            localEntryArray.put(localEntryServiceProxy.getEntry(localEntry.getName()));
+                        } else {
+                            log.error(localEntry.getName() + " was not deployed in the gateway");
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:91]: Better to have this detail in the response

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 6]
```diff
@@ -18,24 +18,101 @@
...
+
+                if (gatewayAPIDTO.getSequenceToBeAdd() != null ) {
+                    JSONArray sequencesArray = new JSONArray();
+                    SequenceAdminServiceProxy sequenceAdminServiceProxy =
+                            new SequenceAdminServiceProxy(gatewayAPIDTO.getTenantDomain());
+                    for (GatewayContentDTO sequence : gatewayAPIDTO.getSequenceToBeAdd()) {
+                        if(sequenceAdminServiceProxy.isExistingSequence(sequence.getName())) {
+                            sequencesArray.put(sequenceAdminServiceProxy.getSequence(sequence.getName()));
+                        } else {
+                            log.error(sequence.getName() + " was not deployed in the gateway");
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:105]: Better to have this detail in the response

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 7]
```diff
@@ -18,24 +18,101 @@
...
+                }
+            } catch (EndpointAdminException e) {
+                log.error("Error in fetching deployed Endpoints from Synapse Configuration." , e);
+            } catch (AxisFault axisFault) {
+                log.error("Error in fetching deployed artifacts from Synapse Configuration." , axisFault);
+            }
+
             responseObj.put("Definition", definition);
             String responseStringObj = String.valueOf(responseObj);
             return Response.ok().entity(responseStringObj).build();
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:118]: we have to handle the errors too. If there was an error, need that details in the response

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/EndPointsApiServiceImpl.java [hunk 8]
```diff
@@ -38,9 +44,19 @@ public Response endPointsGet(String apiName, String label, String apiId, Message
...
+            if (gatewayAPIDTO.getEndpointEntriesToBeAdd() != null ) {
+                EndpointAdminServiceProxy endpointAdminServiceProxy = new EndpointAdminServiceProxy
+                        (gatewayAPIDTO.getTenantDomain());
                 for (GatewayContentDTO gatewayEndpoint : gatewayAPIDTO.getEndpointEntriesToBeAdd()) {
-                    endPointArray.put(gatewayEndpoint.getContent());
+                    try {
+                        if (endpointAdminServiceProxy.isEndpointExist(gatewayEndpoint.getName())) {
+                            endPointArray.put(endpointAdminServiceProxy.getEndpoints(gatewayEndpoint.getName()));
+                        } else {
+                            log.error(gatewayEndpoint.getName() + " was not deployed in the gateway");
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/EndPointsApiServiceImpl.java:55]: better to have this in the response

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/LocalEntryApiServiceImpl.java [hunk 9]
```diff
@@ -39,8 +46,19 @@ public Response localEntryGet(String apiName, String label, String apiId, Messag
...
             if (gatewayAPIDTO.getLocalEntriesToBeAdd() != null) {
+                LocalEntryServiceProxy localEntryServiceProxy = new
+                        LocalEntryServiceProxy(gatewayAPIDTO.getTenantDomain());
                 for (GatewayContentDTO localEntry : gatewayAPIDTO.getLocalEntriesToBeAdd()) {
-                    localEntryArray.put(localEntry.getContent());
+                    try {
+                        if (localEntryServiceProxy.isEntryExists(localEntry.getName())) {
+                            localEntryArray.put(localEntryServiceProxy.getEntry(localEntry.getName()));
+                        } else {
+                            log.error(localEntry.getName() + " was not deployed in the gateway");
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/LocalEntryApiServiceImpl.java:56]: better to have this in the response

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/SequenceApiServiceImpl.java [hunk 10]
```diff
@@ -38,9 +44,19 @@ public Response sequenceGet(String apiName, String label, String apiId, MessageC
...
+            if (gatewayAPIDTO.getSequenceToBeAdd() != null ) {
+                SequenceAdminServiceProxy sequenceAdminServiceProxy =
+                        new SequenceAdminServiceProxy(gatewayAPIDTO.getTenantDomain());
                 for (GatewayContentDTO sequence : gatewayAPIDTO.getSequenceToBeAdd()) {
-                    sequencesArray.put(sequence.getContent());
+                    try {
+                        if (sequenceAdminServiceProxy.isExistingSequence(sequence.getName())) {
+                            sequencesArray.put(sequenceAdminServiceProxy.getSequence(sequence.getName()));
+                        } else {
+                            log.error(sequence.getName() + " was not deployed in the gateway");
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/SequenceApiServiceImpl.java:55]: better to have this in the response

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/resources/gateway-api.yaml [hunk 11]
```diff
@@ -203,9 +203,9 @@ paths:
       security:
         - OAuth2Security:
             - apim:api_publish
-      x-wso2-curl: "curl -k -H \"Authorization: Bearer ae4eae22-3f65-387b-a171-d37eaa366fa8\" -X GET \"https://127.0.0.1:9443/api/am/gateway/v1/api-artifact?apiId=890a4f4d-09eb-4877-a323-57f6ce2ed79b&apiName=PizzaAPI&label=Food\""
+      x-wso2-curl: "curl -k -H \"Authorization: Bearer ae4eae22-3f65-387b-a171-d37eaa366fa8\" -X GET \"https://127.0.0.1:9443/api/am/gateway/v1/api-artifact?apiName=PizzaAPI&version=1.0&tenantDomain=test\""
       x-wso2-request: |
-        GET https://127.0.0.1:9443/api/am/gateway/v1/api-artifact?apiId=890a4f4d-09eb-4877-a323-57f6ce2ed79b&apiName=PizzaAPI&label=Food
+        GET https://127.0.0.1:9443/api/am/gateway/v1/api-artifact?apiName=PizzaAPI&version=1.0&tenantDomain=test
```
Comment (1akshitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/resources/gateway-api.yaml:208]: add a meaningful tenant domain

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/resources/gateway-api.yaml]: https://github.com/wso2/carbon-apimgt/pull/8784/commits/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/SequenceApiServiceImpl.java (hunk 10 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/SequenceApiServiceImpl.java:55]: https://github.com/wso2/carbon-apimgt/pull/8784/commits/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/LocalEntryApiServiceImpl.java (hunk 9 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/LocalEntryApiServiceImpl.java:56]: https://github.com/wso2/carbon-apimgt/pull/8784/commits/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/EndPointsApiServiceImpl.java (hunk 8 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/EndPointsApiServiceImpl.java:55]: https://github.com/wso2/carbon-apimgt/pull/8784/commits/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java (hunk 6 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:113]: https://github.com/wso2/carbon-apimgt/commit/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java (hunk 5 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:96]: https://github.com/wso2/carbon-apimgt/commit/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java (hunk 4 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:79]: https://github.com/wso2/carbon-apimgt/commit/2bef8ac27b023cef0271216e1b2942bc739efacc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java (hunk 2 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java]: https://github.com/wso2/carbon-apimgt/pull/8784/commits/79c1b54ed87ee44cc8828ac3065a843208a9967e

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java (hunk 1 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:47]: https://github.com/wso2/carbon-apimgt/pull/8784/commits/79c1b54ed87ee44cc8828ac3065a843208a9967e

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java (hunk 3 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:62]: This call will only fetch the API Artifacts which have the artifact status of "publish" .  https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/InMemoryAPIDeployer.java#L70

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/gen/java/org/wso2/carbon/apimgt/internal/service/SynapseAttributesApi.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+public class SynapseAttributesApi  {
+
+  @Context MessageContext securityContext;
+
+SynapseAttributesApiService delegate = new SynapseAttributesApiServiceImpl();
+
+
+    @GET
+    
+
```
Comment (chashikajw) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/gen/java/org/wso2/carbon/apimgt/internal/service/SynapseAttributesApi.java:41]: extra line spaces

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 13]
```diff
@@ -246,9 +246,70 @@ public boolean isAPIArtifactExists(String APIId, String gatewayLabel) throws API
         return false;
     }
 
+
+
```
Comment (chashikajw) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:250]: Extra added line spaces

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/gen/java/org/wso2/carbon/apimgt/internal/service/SynapseAttributesApi.java (hunk 12 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/gen/java/org/wso2/carbon/apimgt/internal/service/SynapseAttributesApi.java:41]: That is an Auto-generated File. We don't need to modify anything there.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java [hunk 14]
```diff
@@ -18,25 +18,112 @@
...
+    private static final Log log = LogFactory.getLog(ApiArtifactApiServiceImpl.class);
+    private final String SUPER_TENAT_DOMAIN = "carbon.super";
+
+    @Override
+    public Response apiArtifactGet(String apiName, String version , String tenantDomain,
+            MessageContext messageContext) {
 
         InMemoryAPIDeployer inMemoryApiDeployer = new InMemoryAPIDeployer();
+        if (tenantDomain == null){
+            tenantDomain =SUPER_TENAT_DOMAIN;
```
[the remark this answers is not in the document]
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java:55]: Fixed in the latest version https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.rest.api.gateway.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/gateway/v1/impl/ApiArtifactApiServiceImpl.java#L54

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java (hunk 13 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:250]: Fixed https://github.com/wso2/carbon-apimgt/pull/9546/commits/8254a1cdf158e9bcb28d09f5f5e5823baa79fba7

**Pull request merged**